### PR TITLE
feat(dev-portal): node content component

### PR DESCRIPTION
**Pull request by lottamus**
- Fixes https://github.com/stoplightio/platform-internal/issues/5645
- Adds handler to fetch node content
- Adds React hook for fetching node content

Can see example of this stuff working in our dev portal PR env: https://dev-portal-fuvmsk8zx-stoplight.vercel.app/docs/platform/ZG9jOjQ0-welcome-to-the-stoplight-docs

### CODE REVIEW

File: packages/elements-dev-portal/src/components/NodeContent/index.tsx [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+    const [resolvedUriWithoutAnchor, hash] = resolvedUri.split('#');
+    const edge = node.outbound_edges.find(edge => edge.uri === url || edge.uri === resolvedUriWithoutAnchor);
+
+    if (edge) {
+      return (
+        <Link to={edge.slug} hash={hash}>
+          {children}
+        </Link>
+      );
+    }
```
Comment (marcelltoth) [inline packages/elements-dev-portal/src/components/NodeContent/index.tsx:49]: What is this? 🤔

File: packages/elements-dev-portal/src/handlers/getNodeContent.ts [hunk 2]
```diff
@@ -0,0 +1,34 @@
+import { Node } from '../interfaces/node';
+
+export const getNodeContent = async (
```
Comment (marcelltoth) [inline packages/elements-dev-portal/src/handlers/getNodeContent.ts:3]: In general we prefer colocating related code, rather than having functional folders. As you can probably see we are moving away from having folders like `utils`, `tests`, etc. 

IMO neither this folder or the function add any value, you could just move it into `useGetNodeContent`. Or actually I would consider inlining it, you don't really gain anything by declaring then destructuring parameters again.

File: packages/elements-dev-portal/src/handlers/getNodeContent.ts [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+    throw new Error(data);
+  }
+
+  return data;
+};
+
+function getNodeIdFromSlug(nodeSlug: string) {
+  const [nodeId] = nodeSlug.split('-');
+  return nodeId;
+}
```
Comment (marcelltoth) [inline packages/elements-dev-portal/src/handlers/getNodeContent.ts:34]: Destructuring is nice, but not everything is a nail 😄 

```suggestion
function getNodeIdFromSlug(nodeSlug: string) {
  return nodeSlug.split('-')[0];
}
```

File: packages/elements-dev-portal/src/hooks/useGetNodeContent.ts [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+  {
+    nodeSlug,
+    projectId,
+    branchSlug,
+  }: {
+    nodeSlug: string;
+    projectId: string;
+    branchSlug?: string;
+  },
+  requestHeaders?: Record<string, string>,
```
Comment (marcelltoth) [inline packages/elements-dev-portal/src/hooks/useGetNodeContent.ts:17]: I'm not sure I like this `requestHeaders` parameter. It's leaking the otherwise solid abstraction. Also as the rest are in a nicely structured object, this feels like something that was added last minute, feels like tech debt. What are you trying to achieve with it?

File: packages/elements-dev-portal/src/interfaces/node.ts [hunk 5]
```diff
@@ -0,0 +1,28 @@
+export type Node = NodeSummary & {
```
Comment (marcelltoth) [inline packages/elements-dev-portal/src/interfaces/node.ts:1]: I agree it's kind of hard to colocate this with anything, that's for sure.

But it's certainly not `interfaces`, there isn't a single `interface` in this file. What about doing what we always do, and creating a top level `types.ts`, then forward-exporting the `Node` type from `index.ts`? That is probably the only thing from here that we really need.

File: packages/elements/src/containers/API.tsx [hunk 6]
```diff
@@ -58,11 +56,17 @@ const propsAreWithDocument = (props: APIProps): props is APIPropsWithDocument =>
...
 
   const documentShouldBeFetched = apiDescriptionUrl && !apiDescriptionDocument;
-  const { data: fetchedDocument, error } = useSwr(documentShouldBeFetched ? apiDescriptionUrl! : null, fetcher);
+  const { data: fetchedDocument, error } = useQuery(
+    [apiDescriptionUrl],
+    () => fetch(apiDescriptionUrl).then(res => res.text()),
+    {
+      enabled: !!documentShouldBeFetched,
+    },
+  );
```
Comment (marcelltoth) [inline packages/elements/src/containers/API.tsx:69]: I'm assuming this change will be handled another PR, right? The provider one.

File: packages/elements-dev-portal/src/handlers/getNodeContent.ts (hunk 2 above)
Comment (lottamus) [reply packages/elements-dev-portal/src/handlers/getNodeContent.ts:3]: One use case for keeping them separate is if someone needs to call the handler outside a hook, such as in the NextJS dev-portal: https://github.com/stoplightio/dev-portal/blob/make-it-work/pages/docs/%5BprojectBranchSlug%5D/%5B%5B...nodeSlug%5D%5D.tsx#L61-L66

**Comment by lottamus** [comment]
> I'm also a bit confused. This NodeContent component feels almost like a replacement of Docs. What is the difference? Is Docs still relevant? If not, why not just simply update it? Also this doesn't update StoplightProject itself, that is still using the old API and the old components. Shouldn't that switch to the new NodeContent + hook approach, too?

NodeContent is essentially a replacement for the Docs container, but allows you to pass data in from either the `useGetNodeContent` hook or the `getNodeContent` handler (both which interact with the new HTTP API). 

In an attempt to keep these PRs small (with minimal breaking changes), I've decided to not touch the StoplightProject nor remove any existing code until the end. https://github.com/stoplightio/elements/pull/1195

File: packages/elements-dev-portal/src/handlers/getNodeContent.ts (hunk 2 above)
Comment (marcelltoth) [reply packages/elements-dev-portal/src/handlers/getNodeContent.ts:3]: Oh I see. That's why I said we really need to define the public API before we can reason about the code. Can you add the exports that you meant?

File: packages/elements-dev-portal/src/interfaces/node.ts (hunk 5 above)
Comment (lottamus) [reply packages/elements-dev-portal/src/interfaces/node.ts]: Done in https://github.com/stoplightio/elements/pull/1191/commits/7b06bf468d589f7244993df9366dfe1c15a2f9a8

File: packages/elements-dev-portal/src/handlers/getNodeContent.ts (hunk 2 above)
Comment (lottamus) [reply packages/elements-dev-portal/src/handlers/getNodeContent.ts:3]: Done in https://github.com/stoplightio/elements/pull/1191/commits/a04ea35e0dfaa054470cd9cf1a5e5906ea0774f3

File: packages/elements-dev-portal/src/hooks/useGetNodeContent.ts (hunk 4 above)
Comment (lottamus) [reply packages/elements-dev-portal/src/hooks/useGetNodeContent.ts]: Removed them in https://github.com/stoplightio/elements/pull/1191/commits/811e57e8e70e05040bee14a5d55287c6af516a67

File: packages/elements-dev-portal/src/components/NodeContent/index.tsx (hunk 1 above)
Comment (lottamus) [reply packages/elements-dev-portal/src/components/NodeContent/index.tsx:49]: @marcelltoth we can get the `http_service` from the `http_operation`'s inbound edges. Open to suggestions on how best to pass that down to the elements-core security badge

```ts
const httpServiceNode = httpOperationNode.inbound_edges.find(edge => edge.type === 'http_service');
```

**Comment by philsturgeon** [comment]
As concerns are resolved (questions answered or code changes fix things) can we mark the discussions as resolved so people can read things easier.

File: packages/elements-dev-portal/src/components/NodeContent/index.tsx (hunk 1 above)
Comment (mallachari) [reply packages/elements-dev-portal/src/components/NodeContent/index.tsx:49]: If we can easily get http service uri from operation node, and if we have access to that operation node (I assume we do through context) then we can do the same as we do now, just without the function that @philsturgeon mentioned, which could be removed making things easier.

**Comment by philsturgeon** [comment]
I'll abstain from review on this one and leave it up to the engineers.

**Review by mpodlasin — APPROVED** [review]
LGTM
